File: collatz.jl
collatz(x::Integer) = iseven(x) ? x ÷ 2 : 3x + 1

function collatz_sequence(x::Integer)
    s = [x]
    while x > 1
        x = collatz(x)
        push!(s, x)
    end
    return s
end

function benchmark(N=1000)
    return [length(collatz_sequence(x)) for x in range(1, N)]
end

# using BenchmarkTools
# @btime benchmark()
# 251.833 μs (3531 allocations: 1.24 MiB)

File: collatz.R
collatz = function(x) {
    ifelse(x %% 2 == 0, x %/% 2, 3 * x + 1)
}

collatz_sequence = function(x) {
    s = x
    while (x > 1) {
        x = collatz(x)
        s = c(s, x)
    }
    s
}

benchmark = function(N=1000) {
    s = integer(N)
    for (x in 1:N) {
        s[x] = length(collatz_sequence(x))
    }
    s
}

# microbenchmark::microbenchmark(benchmark())
# Unit: milliseconds
#        expr      min       lq     mean   median       uq      max neval
# benchmark() 67.83118 68.57992 70.35936 69.13508 70.23226 97.84818   100
